modules/delegates.js

```diff
@@ -429,69 +423,68 @@ __private.checkDelegates = function(senderPublicKey, votes, state, cb, tx) {
 
 						if (
 							!votesAccounts ||
-							votesAccounts.length < voteAddressesWithActions.length
+							votesAccounts.length < votesWithAction.length
 						) {
 							library.logger.error(
 								'Delegates with addresses not found',
 								_.differenceWith(
-									voteAddressesWithActions,
+									votesWithAction,
 									votesAccounts,
 									(
-										{ address: addressFromTransaction },
-										{ addressFromAccount }
-									) => addressFromTransaction === addressFromAccount
-								).map(({ address }) => address)
+										{ publicKey: publicKeyFromTransaction },
+										{ publicKey: publicKeyFromAccount }
+									) => publicKeyFromTransaction === publicKeyFromAccount
+								).map(({ publicKey }) => publicKey)
 							);
 							return setImmediate(waterfallCb, 'Delegate not found');
 						}
-
 						const upvoteAccounts = votesAccounts.filter(voteAccount =>
-							voteAddressesWithActions.find(
-								actionWithAddress =>
-									actionWithAddress.action === '+' &&
-									actionWithAddress.address === voteAccount.address
+							votesWithAction.find(
+								({ action, publicKey }) =>
+									action === '+' && publicKey === voteAccount.publicKey
 							)
 						);
```

Comment:
This will be much simpler. 

```js
const upVotePublicKeys  = votesWithAction.filter(({ action }) => action === '+').map(({publicKey}) => publicKey);
```

Reply:
We need upvoteAccounts, not upvotePublicKeys because we need to return an error with the username of the account if the account has already voted for the delegate.